### PR TITLE
Fix #48: correctly handle expansion of hidden nested iron-collapse

### LINKED ISSUE #48: Programmatically calling show() on a hidden submenu puts it in a buggy state
### Description

Programmatically expanding (`show()`) a hidden submenu puts it in a buggy state, where `opened` is `true` but the CSS `height` is `0px`.

This is due to a bug in the animation code, and therefore only manifests if the submenu has animations turned on (which is the default).

### Live Demo

http://jsbin.com/rutakeciri/edit?html,output

### Steps to reproduce

1. Click on the button (calls `show()` on the closed `iron-collapse` nested under a closed top-level `iron-collapse`).
2. Click on the top-level `iron-collapse`.

### Expected outcome

The top-level `iron-collapse` opens and reveals an open nested `iron-collapse`.

### Actual outcome

The top-level `iron-collapse` opens and reveals an apparently closed nested `iron-collapse`. (In fact its `opened` is `true` but its `height` is `0`.

Clicking on the nested `iron-collapse` seemingly does nothing (in fact it sets `opened` to `false`).

Clicking it *again* reveals it as expected.

This is due to the way [the animation code](https://github.com/PolymerElements/iron-collapse/blob/master/iron-collapse.html#L178) tries to calculate the target height by setting the desired style and calling `getBoundingClientRect`. Unfortunately when the element is nested under a hidden parent menu (`display: none`), gBCR returns 0 as height.

### Browsers Affected

- [x] Chrome
- [x] Firefox
- [ ] Safari 9
- [ ] Safari 8
- [ ] Safari 7
- [ ] Edge
- [ ] IE 11
- [ ] IE 10

See https://github.com/PolymerElements/iron-collapse/pull/49 for a fix.

Note that I've encountered this issue when using `paper-menu` (which uses `iron-collapse` internally).

### PR DESCRIPTION
**Pull request by theefer**
Fixes https://github.com/PolymerElements/iron-collapse/issues/48 by disabling the problematic animation code when the element is hidden.

Note that I've used a fairly silly heuristic to determine whether the current element is rendered or not. If there's a better way to do this, please let me know!

### CODE REVIEW

**Comment by theefer** [comment]
Re CLA: Woops, now fixed my commit to use my `@google.com` email.

**Comment by valdrinkoshi** [comment]
Hi @theefer, thanks for the contribution! Could you please add unit tests so we avoid any future regression? Thanks!

File: iron-collapse.html [hunk 1]
```diff
@@ -239,6 +239,14 @@
       this.notifyResize();
     },
 
+    _isVisible: function() {
+      var rect = this.getBoundingClientRect();
```
Comment (valdrinkoshi) [inline iron-collapse.html]: this could just be
```javascript
return this.getBoundingClientRect()[this.dimension] > 0;

**Comment by valdrinkoshi** [comment]
Also, please follow our [Contributing guidelines](https://github.com/PolymerElements/iron-collapse/blob/master/CONTRIBUTING.md#submitting-pull-requests)

File: iron-collapse.html (hunk 1 above)
Comment (theefer) [inline iron-collapse.html]: I don't think this would work, because `_isVisible` is called on the pristine element, before its `style[this.dimension]` is set to the target size. That means that `_isVisible` would return `false` for say a top-level `iron-collapse` component that is displayed but closed; we don't want to skip the animation in that case.

In other words, `visible` here means "is not `display: none`", which is different from the `opened` state.

**Comment by theefer** [comment]
@valdrinkoshi Thanks, I've added some tests now. I didn't squash the commit to make it easier to see my change since your comment, and I believe GitHub now allows you to squash when you merge the branch?

Let me know if you need any more changes.

File: iron-collapse.html (hunk 1 above)
Comment (valdrinkoshi) [inline iron-collapse.html]: Yes, but the suggested version still checks the bounding rect, it just limit to check 1 dimension (width or height, see the getter of `dimension`), while yours checks for `left, right, top, bottom` too. This should not return false if top = 0

Comment (valdrinkoshi) [inline iron-collapse.html]: Sorry, I now realize it checks for anything to be non-zero. This works too

File: test/nested.html [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            assert.equal(innerCollapse.style.transitionDuration, '0s');
+            assert.equal(innerCollapse.style.height, 'auto');
+          });
+
+        });
+
+        suite('opened outer collapse', function() {
+
+          setup(function () {
+            outerCollapse.opened = true;
```
Comment (valdrinkoshi) [inline test/nested.html]: In the reported issue we don't start with an opened outerCollapse, we first open the innerCollapse, then the outerCollapse.

File: test/nested.html [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+          test('inner collapse default opened attribute', function() {
+            assert.equal(innerCollapse.opened, false);
+          });
+
+          test('inner collapse default style height', function() {
+            assert.equal(innerCollapse.style.height, '0px');
+          });
+
+          test('set inner opened to true triggers height animation', function() {
+            innerCollapse.opened = true;
```
Comment (valdrinkoshi) [inline test/nested.html]: This test should open inner, then outer, then check if inner bounding rect has a proper height.

Comment (valdrinkoshi) [inline test/nested.html]: Also, can you write this same test for a horizontal innerCollapse & check it has a non-zero width?

**Comment by valdrinkoshi** [comment]
@theefer I've let some comments ;)

File: iron-collapse.html [hunk 4]
```diff
@@ -239,6 +239,14 @@
       this.notifyResize();
     },
 
+    _isVisible: function() {
```
Comment (valdrinkoshi) [inline iron-collapse.html]: Since this method tells us if is displayed, can we name it `_isDisplayed`?
Also, any reason why making this a method and not a getter?
Finally, can you add a jsdoc for it & mark it `@private`?

File: test/nested.html (hunk 3 above)
Comment (theefer) [inline test/nested.html]: Added that too.

File: test/nested.html [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            Lorem ipsum
+          </div>
+
+          <iron-collapse id="inner-collapse-vertical">
+            <div style="height:100px;">
+              consectetur adipiscing elit
+            </div>
+          </iron-collapse>
+
+          <iron-collapse id="inner-collapse-horizontal" horizontal style="display: inline-block">
```
Comment (valdrinkoshi) [inline test/nested.html]: no action required, but I now realize that the `iron-collapse` should set `display: inline-block` when `horizontal`...Will open an issue to keep track of this!

Comment (valdrinkoshi) [inline test/nested.html]: Actually, that's not needed, we can set the `width` even for `display: block`...
Why is this style needed here?

File: iron-collapse.html (hunk 4 above)
Comment (theefer) [inline iron-collapse.html]: Done (though can't find other iron elements that use `@private` jsdoc?)

File: test/nested.html (hunk 5 above)
Comment (theefer) [inline test/nested.html]: Cool, I think I noticed it had a small visible impact on the animation if the content's width is much smaller than the `iron-collapse` width (which fills its own parent).

Comment (valdrinkoshi) [inline test/nested.html]: Oh I see..thanks!

**Comment by valdrinkoshi** [comment]
LGTM, good job @theefer :+1:

**Comment by theefer** [comment]
Thanks for the review.

Out of interest, can this go out as a quick patch release or does it have to wait on a wider PolymerElements release cycle? (I initially dug into this issue as it is blocking a fix to an internal project.)

**Comment by valdrinkoshi** [comment]
Yes, I'll release a new release as soon as tests on master are green as well

**Comment by theefer** [comment]
Amazing, thanks @valdrinkoshi!

**Comment by valdrinkoshi** [comment]
All green! Released `v1.0.7` ;)
